### PR TITLE
audio_core: Dynamically load libfdk-aac if available.

### PR DESCRIPTION
**Pull request by Steveice10**
Introduces a cross-platform dynamic library loading utility, and uses it to dynamically load libfdk-aac at runtime instead of directly linking. This allows us to support using libfdk-aac only when the user provides it themselves, and falling back to the null decoder otherwise.

As part of this change the way decoders are selected has also been revised so that all decoders supported by a build can be considered, rather than just one.

- - -
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/6568)

### CODE REVIEW

File: src/audio_core/hle/fdk_decoder.cpp [hunk 1]
```diff
@@ -2,11 +2,110 @@
...
+};
+
+struct LIB_INFO {
+    const char* title;
+    const char* build_date;
+    const char* build_time;
+    FDK_MODULE_ID module_id;
+    s32 version;
+    u32 flags;
+    char versionStr[32];
```
Comment (FearlessTobi) [inline src/audio_core/hle/fdk_decoder.cpp:40]:
```suggestion
    std::array<char, 32> version_string;
```

File: src/audio_core/hle/fdk_decoder.cpp [hunk 2]
```diff
@@ -2,11 +2,110 @@
...
+enum AAC_DECODER_ERROR {
+    AAC_DEC_OK = 0,
+    AAC_DEC_TRANSPORT_SYNC_ERROR = 0x1001,
+};
+
+enum AACDEC_PARAM {
+    AAC_PCM_MAX_OUTPUT_CHANNELS = 0x12,
+};
+
+struct LIB_INFO {
```
Comment (FearlessTobi) [inline src/audio_core/hle/fdk_decoder.cpp:33]: those struct should probably have a static_assert for their size

File: src/common/dynamic_library.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+#include <windows.h>
+#else
+#include <dlfcn.h>
+#endif
+
+class DynamicLibrary {
+public:
+    DynamicLibrary(const std::string& name) {
+#if defined(_WIN32)
+        auto full_name = name + ".dll";
```
Comment (FearlessTobi) [inline src/common/dynamic_library.h:17]:
```suggestion
        const auto full_name = name + ".dll";
```
dito below

File: src/common/dynamic_library.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+#if defined(__APPLE__)
+        auto full_name = "lib" + name + ".dylib";
+#else
+        auto full_name = "lib" + name + ".so";
+#endif
+        handle = dlopen(full_name.c_str(), RTLD_LAZY);
+        if (!handle) {
+            load_error = dlerror();
+        }
+#endif
```
Comment (SachinVin) [inline src/common/dynamic_library.h:40]:
```suggestion
#endif // defined(_WIN32)
```

**Comment by Steveice10** [comment]
> Wouldn't it be better to add the fdk headers in externals?

I’m not sure we can actually use the headers without also pulling in the original function definitions which can’t be used and would conflict.

File: src/audio_core/hle/fdk_decoder.cpp (hunk 1 above)
Comment (Steveice10) [reply src/audio_core/hle/fdk_decoder.cpp:40]: Do we want to customize the types and names like this? I had a goal to make this basically just a drop-in replacement where none of the logic had to be tweaked and it works the same as normal fdk-aac usage, but I'm open to tweaking it more if preferred.

**Comment by Steveice10** [comment]
Update on that, I actually gave it a shot and turns out that I can just clobber the original function definitions with the dynamic versions fine, so it's still drop-in. Have changed over to using the original headers for type definitions.

File: externals/fdk-aac/README.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
+# fdk-aac
+
```
Comment (SachinVin) [inline externals/fdk-aac/README.md:2]: Nit, could you add the repo url and commit hash

**Comment by Steveice10** [comment]
Superceded by https://github.com/citra-emu/citra/pull/6570

**Pull request closed without merging**